bmc/oem/dell.go

```diff
@@ -60,3 +65,370 @@ func (r *Dell) GetTaskMonitorDetails(ctx context.Context, taskMonitorResponse *h
+			// as the settings provided will have actual values.
+			// replace display values with actual values
+			if strings.ToLower(string(entry.Type)) == string(redfish.EnumerationAttributeType) {
+				for _, attrValue := range entry.Value {
+					if attrValue.ValueDisplayName == mergedBMCAttributes[name] {
+						result[name] = attrValue.ValueName
+						break
+					}
+				}
+				if _, ok := result[name]; !ok {
+					errs = append(
+						errs,
+						fmt.Errorf(
+							"current setting '%v' for key '%v' not found in possible values for it (%v)",
+							mergedBMCAttributes[name],
+							name,
+							entry.Value,
+						))
+				}
+			} else {
+				result[name] = mergedBMCAttributes[name]
+			}
+		} else {
+			// possible error in settings key
+			errs = append(errs, fmt.Errorf("setting key '%v' not found in possible settings", name))
+		}
+	}
+	if len(errs) > 0 {
+		return result, fmt.Errorf(
+			"some errors found in the settings '%v'.\nPossible settings %v",
+			errs,
+			maps.Keys(filteredAttr),
+		)
+	}
+
+	return result, nil
+}
+
+func (d *DellIdracManager) UpdateBMCAttributesApplyAt(
+	attrs redfish.SettingsAttributes,
+	applyTime common.ApplyTime,
+) error {
+
+	BMCattributeValues, err := d.getCurrentBMCSettingAttribute()
+	if err != nil {
+		return err
+	}
+
+	payloads := make(map[string]redfish.SettingsAttributes, len(BMCattributeValues))
+	for key, value := range attrs {
+		for _, eachAttr := range BMCattributeValues {
+			if _, ok := eachAttr.Attributes[key]; ok {
+				if data, ok := payloads[eachAttr.Settings.SettingsObject.String()]; ok {
+					data[key] = value
+				} else {
+					payloads[eachAttr.Settings.SettingsObject.String()] = make(redfish.SettingsAttributes)
+					payloads[eachAttr.Settings.SettingsObject.String()][key] = value
+				}
+				// keys cant be duplicate. Hence, break once its already found in one of idrac settings sub types
+				break
+			}
+		}
+	}
+
+	// If there are any allowed updates, try to send updates to the system and
+	// return the result.
+	if len(payloads) > 0 {
+		var errs []error
+		// for wach sub type, apply the settings
+		for settingPath, payload := range payloads {
+			// fetch the etag required for settingPath
+			etag, err := func(uri string) ([]string, error) {
+				resp, err := d.BMC.GetClient().Get(uri)
+				if err != nil {
+					return nil, err
+				}
+				defer resp.Body.Close() // nolint: errcheck
+				return resp.Header["Etag"], nil
+			}(settingPath)
+
+			if err != nil {
+				errs = append(errs, fmt.Errorf("failed to get Etag for %v. error %v", settingPath, err))
+				continue
+			}
+
+			data := map[string]interface{}{"Attributes": payload}
+			if applyTime != "" {
+				data["@Redfish.SettingsApplyTime"] = map[string]string{"ApplyTime": string(applyTime)}
+			}
+			var header = make(map[string]string)
+			if etag != nil {
+				header["If-Match"] = etag[0]
+			}
+
+			err = func(uri string, data map[string]any, header map[string]string) error {
```

Comment:
Function arguments are not needed here. Local variables can be directly used in a closure.